server/router/src/main/java/org/infinispan/server/router/handlers/SniHandlerInitializer.java

```diff
@@ -0,0 +1,43 @@
+package org.infinispan.server.router.handlers;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelInitializer;
+import io.netty.handler.ssl.SslContext;
+import io.netty.util.DomainMappingBuilder;
+import io.netty.util.DomainNameMapping;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.server.router.RoutingTable;
+import org.infinispan.server.router.handlers.util.SslUtils;
+import org.infinispan.server.router.logging.RouterLogger;
+import org.infinispan.server.router.routes.RouteDestination;
+import org.infinispan.server.router.routes.SniRouteSource;
+
+import javax.net.ssl.SSLContext;
+import java.lang.invoke.MethodHandles;
+
+public class SniHandlerInitializer extends ChannelInitializer<Channel> {
+
+    private static final RouterLogger logger = LogFactory.getLog(MethodHandles.lookup().lookupClass(), RouterLogger.class);
+
+    private final RoutingTable routingTable;
+
+    public SniHandlerInitializer(RoutingTable routingTable) {
+        this.routingTable = routingTable;
+    }
+
+    @Override
+    protected void initChannel(Channel channel) throws Exception {
+        SslContext defaultContext = SslUtils.INSTANCE.toNettySslContext(SSLContext.getDefault());
+        DomainMappingBuilder<SslContext> domainMappingBuilder = new DomainMappingBuilder<>(defaultContext);
+
+        routingTable.streamRoutes(SniRouteSource.class, RouteDestination.class)
+                .map(r -> r.getRouteSource())
+                .forEach(r -> domainMappingBuilder.add(r.getSniHostName(), SslUtils.INSTANCE.toNettySslContext(r.getSslContext())));
+
+        DomainNameMapping<SslContext> domainNameMapping = domainMappingBuilder.build();
```

Comment:
Sorry, I made this comment too early... no we can not.. for some reason Netty started closing connections (possible some state is written in static fields). Reverting back to `initChannel` unfortunately :(